addons/survey/models/survey.py

```diff
@@ -0,0 +1,1077 @@
+        for survey in self:
+            survey.public_url_html = '<a href="%s">%s</a>' % (survey.public_url, _("Click here to start survey"))
+
+    @api.multi
+    def _compute_print_url(self):
+        """ Computes a printing URL for the survey """
+        base_url = '/' if self.env.context.get('relative_url') else self.env['ir.config_parameter'].get_param('web.base.url')
+        for survey in self:
+            survey.print_url = urljoin(base_url, "survey/print/%s" % slug(survey))
+
+    @api.multi
+    def _compute_result_url(self):
+        """ Computes an URL for the survey results """
+        base_url = '/' if self.env.context.get('relative_url') else self.env['ir.config_parameter'].get_param('web.base.url')
+        for survey in self:
+            survey.result_url = urljoin(base_url, "survey/results/%s" % slug(survey))
+
+    @api.multi
+    def _read_group_stage_ids(self, domain, read_group_order=None, access_rights_uid=None):
+        """ Read group customization in order to display all the stages in the
+        kanban view, even if they are empty """
+        SurveyStage = self.env['survey.stage']
+        order = SurveyStage._order
+        access_rights_uid = access_rights_uid or self.env.user.id
+
+        if read_group_order == 'stage_id desc':
+            order = '%s desc' % order
+
+        stage_ids = SurveyStage._search([], order=order, access_rights_uid=access_rights_uid)
+        stage = SurveyStage.browse(stage_ids)
+        result = stage.sudo(access_rights_uid).name_get()
+
+        # restore order of the search
+        result.sort(lambda x, y: cmp(stage_ids.index(x[0]), stage_ids.index(y[0])))
+
+        fold = {}
+        for stage in SurveyStage.sudo(access_rights_uid).browse(stage_ids):
+            fold[stage.id] = stage.fold
+        return result, fold
+
+    _group_by_full = {
+        'stage_id': _read_group_stage_ids
+    }
+
+    # Actions
+
+    @api.multi
+    def action_start_survey(self):
+        ''' Open the website page with the survey form '''
+        self.ensure_one()
+        trail = ""
+        context = dict(self.env.context, relative_url=True)
+        if 'survey_token' in context:
+            trail = "/" + context['survey_token']
+        return {
+            'type': 'ir.actions.act_url',
+            'name': "Start Survey",
+            'target': 'self',
+            'url': self.with_context(context).public_url + trail
+        }
+
+    @api.multi
+    def action_send_survey(self):
+        ''' Open a window to compose an email, pre-filled with the survey
+        message '''
+        self.ensure_one()
+        if not self.designed:
+            raise UserError(_('You cannot send an invitation for a survey that has no questions.'))
+
+        if self.stage_id.closed:
+            raise UserError(_("You cannot send invitations for closed surveys."))
+        template_id = self.env.ref('survey.email_template_survey').id
```

Comment:
What if email_template_survey not found, should it raise ValueError or it should silently send default_template_id = False(I would prefer later one)